// cdk-assets/lib/private/placeholders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk_assets_schema_1 = require("@aws-cdk/cdk-assets-schema");
/**
 * Replace the {ACCOUNT} and {REGION} placeholders in all strings found in a complex object.
 *
 * Duplicated between cdk-assets and aws-cdk CLI because we don't have a good single place to put it
 * (they're nominally independent tools).
 */
async function replaceAwsPlaceholders(object, aws) {
    let region;
    let account;
    return await recurse(object);
    async function recurse(value) {
        var _a, _b;
        if (typeof value === 'string') {
            if (value.indexOf(cdk_assets_schema_1.Placeholders.CURRENT_REGION) > -1) {
                await ensureRegion();
            }
            if (value.indexOf(cdk_assets_schema_1.Placeholders.CURRENT_ACCOUNT) > -1) {
                await ensureAccount();
            }
            if (value.indexOf(cdk_assets_schema_1.Placeholders.CURRENT_PARTITION) > -1) {
                await ensureAccount();
            }
            value = replaceAll(value, cdk_assets_schema_1.Placeholders.CURRENT_REGION, region !== null && region !== void 0 ? region : 'WONTHAPPEN');
            value = replaceAll(value, cdk_assets_schema_1.Placeholders.CURRENT_ACCOUNT, (_a = account === null || account === void 0 ? void 0 : account.accountId) !== null && _a !== void 0 ? _a : 'WONTHAPPEN');
            value = replaceAll(value, cdk_assets_schema_1.Placeholders.CURRENT_PARTITION, (_b = account === null || account === void 0 ? void 0 : account.partition) !== null && _b !== void 0 ? _b : 'WONTHAPPEN');
            return value;
        }
        if (typeof value !== 'object' || value === null) {
            return value;
        }
        if (Array.isArray(value)) {
            return value.map(recurse);
        }
        const ret = {};
        for (const [key, inner] of Object.entries(value)) {
            ret[key] = await recurse(inner);
        }
        return ret;
    }
    async function ensureRegion() {
        var _a;
        if (region === undefined) {
            region = (_a = object.region) !== null && _a !== void 0 ? _a : await aws.discoverDefaultRegion();
        }
    }
    async function ensureAccount() {
        if (account === undefined) {
            account = await aws.discoverCurrentAccount();
        }
    }
}
exports.replaceAwsPlaceholders = replaceAwsPlaceholders;
/**
 * A "replace-all" function that doesn't require us escaping a literal string to a regex
 */
function replaceAll(s, search, replace) {
    return s.split(search).join(replace);
}
