// @aws-cdk/aws-cognito/test/integ.user-pool-explicit-props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const aws_lambda_1 = require("@aws-cdk/aws-lambda");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
const app = new core_1.App();
const stack = new core_1.Stack(app, 'integ-user-pool');
const userpool = new lib_1.UserPool(stack, 'myuserpool', {
    userPoolName: 'MyUserPool',
    userInvitation: {
        emailSubject: 'invitation email subject from the integ test',
        emailBody: 'invitation email body from the integ test for {username}. Temp password is {####}.',
        smsMessage: 'invitation sms message from the integ test for {username}. Temp password is {####}.',
    },
    selfSignUpEnabled: true,
    userVerification: {
        emailBody: 'verification email body from the integ test. Code is {####}.',
        emailSubject: 'verification email subject from the integ test',
        smsMessage: 'verification sms message from the integ test. Code is {####}.',
    },
    signInAliases: {
        username: true,
        email: true,
    },
    autoVerify: {
        email: true,
        phone: true,
    },
    requiredAttributes: {
        fullname: true,
        email: true,
    },
    customAttributes: {
        'some-string-attr': new lib_1.StringAttribute(),
        'another-string-attr': new lib_1.StringAttribute({ minLen: 4, maxLen: 100 }),
        'some-number-attr': new lib_1.NumberAttribute(),
        'another-number-attr': new lib_1.NumberAttribute({ min: 10, max: 50 }),
        'some-boolean-attr': new lib_1.BooleanAttribute(),
        'some-datetime-attr': new lib_1.DateTimeAttribute(),
    },
    mfa: lib_1.Mfa.REQUIRED,
    mfaSecondFactor: {
        sms: true,
        otp: true,
    },
    passwordPolicy: {
        tempPasswordValidity: core_1.Duration.days(10),
        minLength: 12,
        requireDigits: true,
        requireLowercase: true,
        requireUppercase: true,
        requireSymbols: true,
    },
    emailSettings: {
        from: 'noreply@myawesomeapp.com',
        replyTo: 'support@myawesomeapp.com',
    },
    lambdaTriggers: {
        createAuthChallenge: dummyTrigger('createAuthChallenge'),
        customMessage: dummyTrigger('customMessage'),
        defineAuthChallenge: dummyTrigger('defineAuthChallenge'),
        postAuthentication: dummyTrigger('postAuthentication'),
        postConfirmation: dummyTrigger('postConfirmation'),
        preAuthentication: dummyTrigger('preAuthentication'),
        preSignUp: dummyTrigger('preSignUp'),
        preTokenGeneration: dummyTrigger('preTokenGeneration'),
        userMigration: dummyTrigger('userMigration'),
        verifyAuthChallengeResponse: dummyTrigger('verifyAuthChallengeResponse'),
    },
});
const cognitoDomain = userpool.addDomain('myuserpooldomain', {
    cognitoDomain: {
        domainPrefix: 'myawesomeapp',
    },
});
new core_1.CfnOutput(stack, 'userpoolId', {
    value: userpool.userPoolId,
});
new core_1.CfnOutput(stack, 'cognitoDomainName', {
    value: `${cognitoDomain.domainName}.auth.${stack.region}.amazoncognito.com`,
});
function dummyTrigger(name) {
    return new aws_lambda_1.Function(stack, name, {
        functionName: name,
        handler: 'index.handler',
        runtime: aws_lambda_1.Runtime.NODEJS_12_X,
        code: aws_lambda_1.Code.fromInline('foo'),
    });
}
