// @aws-cdk/aws-cognito/test/integ.user-pool-client-explicit-props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
const app = new core_1.App();
const stack = new core_1.Stack(app, 'integ-user-pool-client-explicit-props');
const userpool = new lib_1.UserPool(stack, 'myuserpool');
userpool.addClient('myuserpoolclient', {
    userPoolClientName: 'myuserpoolclient',
    authFlows: {
        adminUserPassword: true,
        custom: true,
        refreshToken: true,
        userPassword: true,
        userSrp: true,
    },
    generateSecret: true,
    oAuth: {
        flows: {
            implicitCodeGrant: true,
            authorizationCodeGrant: true,
        },
        scopes: [
            lib_1.OAuthScope.PHONE,
            lib_1.OAuthScope.EMAIL,
            lib_1.OAuthScope.OPENID,
            lib_1.OAuthScope.PROFILE,
            lib_1.OAuthScope.COGNITO_ADMIN,
            lib_1.OAuthScope.custom('my-resource-server/my-scope'),
        ],
        callbackUrls: ['https://redirect-here.myapp.com'],
    },
});
