// aws-cdk/lib/api/aws-auth/sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AWS = require("aws-sdk");
const logging_1 = require("../../logging");
const functions_1 = require("../../util/functions");
const account_cache_1 = require("./account-cache");
/**
 * Base functionality of SDK without credential fetching
 */
class SDK {
    constructor(credentials, region, httpOptions = {}) {
        this.credentials = credentials;
        /**
         * Default retry options for SDK clients
         *
         * Biggest bottleneck is CloudFormation, with a 1tps call rate. We want to be
         * a little more tenacious than the defaults, and with a little more breathing
         * room between calls (defaults are {retries=3, base=100}).
         *
         * I've left this running in a tight loop for an hour and the throttle errors
         * haven't escaped the retry mechanism.
         */
        this.retryOptions = { maxRetries: 6, retryDelayOptions: { base: 300 } };
        this.config = {
            ...httpOptions,
            ...this.retryOptions,
            credentials,
            region,
        };
        this.currentRegion = region;
    }
    cloudFormation() {
        return new AWS.CloudFormation(this.config);
    }
    ec2() {
        return new AWS.EC2(this.config);
    }
    ssm() {
        return new AWS.SSM(this.config);
    }
    s3() {
        return new AWS.S3(this.config);
    }
    route53() {
        return new AWS.Route53(this.config);
    }
    ecr() {
        return new AWS.ECR(this.config);
    }
    async currentAccount() {
        return functions_1.cached(this, CURRENT_ACCOUNT_KEY, () => SDK.accountCache.fetch(this.credentials.accessKeyId, async () => {
            // if we don't have one, resolve from STS and store in cache.
            logging_1.debug('Looking up default account ID from STS');
            const result = await new AWS.STS(this.config).getCallerIdentity().promise();
            const accountId = result.Account;
            const partition = result.Arn.split(':')[1];
            if (!accountId) {
                throw new Error('STS didn\'t return an account ID');
            }
            logging_1.debug('Default account ID:', accountId);
            return { accountId, partition };
        }));
    }
}
exports.SDK = SDK;
SDK.accountCache = new account_cache_1.AccountAccessKeyCache();
const CURRENT_ACCOUNT_KEY = Symbol('current_account_key');
