// @aws-cdk/assert/lib/assertion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Assertion {
    /**
     * Assert this thing and another thing
     */
    and(assertion) {
        // Needs to delegate to a function so that we can import mutually dependent classes in the right order
        return and(this, assertion);
    }
    assertOrThrow(inspector) {
        if (!this.assertUsing(inspector)) {
            throw new Error(`${JSON.stringify(inspector.value, null, 2)} does not match ${this.description}`);
        }
    }
}
exports.Assertion = Assertion;
class JestFriendlyAssertion extends Assertion {
}
exports.JestFriendlyAssertion = JestFriendlyAssertion;
const and_assertion_1 = require("./assertions/and-assertion");
function and(left, right) {
    return new and_assertion_1.AndAssertion(left, right);
}
const negated_assertion_1 = require("./assertions/negated-assertion");
function not(assertion) {
    return new negated_assertion_1.NegatedAssertion(assertion);
}
exports.not = not;
