// aws-cdk/lib/api/bootstrap/deploy-bootstrap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cxschema = require("@aws-cdk/cloud-assembly-schema");
const cxapi = require("@aws-cdk/cx-api");
const fs = require("fs-extra");
const os = require("os");
const path = require("path");
const aws_auth_1 = require("../aws-auth");
const deploy_stack_1 = require("../deploy-stack");
const toolkit_info_1 = require("../toolkit-info");
const bootstrap_props_1 = require("./bootstrap-props");
/**
 * Perform the actual deployment of a bootstrap stack, given a template and some parameters
 */
async function deployBootstrapStack(template, parameters, environment, sdkProvider, options) {
    var _a, _b, _c;
    const toolkitStackName = (_a = options.toolkitStackName) !== null && _a !== void 0 ? _a : toolkit_info_1.DEFAULT_TOOLKIT_STACK_NAME;
    const resolvedEnvironment = await sdkProvider.resolveEnvironment(environment);
    const sdk = await sdkProvider.forEnvironment(resolvedEnvironment, aws_auth_1.Mode.ForWriting);
    const newVersion = bootstrapVersionFromTemplate(template);
    const currentBootstrapStack = await toolkit_info_1.ToolkitInfo.lookup(resolvedEnvironment, sdk, toolkitStackName);
    if (currentBootstrapStack && newVersion < currentBootstrapStack.version && !options.force) {
        throw new Error(`Not downgrading existing bootstrap stack from version '${currentBootstrapStack.version}' to version '${newVersion}'. Use --force to force.`);
    }
    const outdir = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk-bootstrap'));
    const builder = new cxapi.CloudAssemblyBuilder(outdir);
    const templateFile = `${toolkitStackName}.template.json`;
    await fs.writeJson(path.join(builder.outdir, templateFile), template, { spaces: 2 });
    builder.addArtifact(toolkitStackName, {
        type: cxschema.ArtifactType.AWS_CLOUDFORMATION_STACK,
        environment: cxapi.EnvironmentUtils.format(environment.account, environment.region),
        properties: {
            templateFile,
        },
    });
    const assembly = builder.buildAssembly();
    return await deploy_stack_1.deployStack({
        stack: assembly.getStackByName(toolkitStackName),
        resolvedEnvironment,
        sdk: await sdkProvider.forEnvironment(resolvedEnvironment, aws_auth_1.Mode.ForWriting),
        sdkProvider,
        roleArn: options.roleArn,
        tags: (_b = options.parameters) === null || _b === void 0 ? void 0 : _b.tags,
        execute: (_c = options === null || options === void 0 ? void 0 : options.parameters) === null || _c === void 0 ? void 0 : _c.execute,
        parameters,
    });
}
exports.deployBootstrapStack = deployBootstrapStack;
function bootstrapVersionFromTemplate(template) {
    var _a, _b, _c;
    return parseInt((_c = (_b = (_a = template.Outputs) === null || _a === void 0 ? void 0 : _a[bootstrap_props_1.BOOTSTRAP_VERSION_OUTPUT]) === null || _b === void 0 ? void 0 : _b.Value) !== null && _c !== void 0 ? _c : '0', 10);
}
