// aws-cdk/test/api/bootstrap2.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mockDeployStack = jest.fn();
jest.mock('../../lib/api/deploy-stack', () => ({
    deployStack: mockDeployStack,
}));
let mockToolkitInfo;
jest.mock('../../lib/api/toolkit-info', () => ({
    // Pretend there's no toolkit deployed yet
    DEFAULT_TOOLKIT_STACK_NAME: 'CDKToolkit',
    ToolkitInfo: {
        lookup: () => mockToolkitInfo,
    },
}));
const bootstrap_1 = require("../../lib/api/bootstrap");
const mock_sdk_1 = require("../util/mock-sdk");
describe('Bootstrapping v2', () => {
    const env = {
        account: '123456789012',
        region: 'us-east-1',
        name: 'mock',
    };
    const sdk = new mock_sdk_1.MockSdkProvider();
    mockToolkitInfo = undefined;
    test('passes the bucket name as a CFN parameter', async () => {
        await bootstrap_1.bootstrapEnvironment2(env, sdk, {
            parameters: {
                bucketName: 'my-bucket-name',
            },
        });
        expect(mockDeployStack).toHaveBeenCalledWith(expect.objectContaining({
            parameters: {
                FileAssetsBucketName: 'my-bucket-name',
            },
        }));
    });
    test('passes the KMS key ID as a CFN parameter', async () => {
        await bootstrap_1.bootstrapEnvironment2(env, sdk, {
            parameters: {
                kmsKeyId: 'my-kms-key-id',
            },
        });
        expect(mockDeployStack).toHaveBeenCalledWith(expect.objectContaining({
            parameters: {
                FileAssetsBucketKmsKeyId: 'my-kms-key-id',
            },
        }));
    });
    test('passing trusted accounts without CFN managed policies results in an error', async () => {
        await expect(bootstrap_1.bootstrapEnvironment2(env, sdk, {
            parameters: {
                trustedAccounts: ['123456789012'],
            },
        }))
            .rejects
            .toThrow('--cloudformation-execution-policies are required if --trust has been passed!');
    });
    test('Do not allow downgrading bootstrap stack version', async () => {
        // GIVEN
        mockToolkitInfo = {
            version: 999,
        };
        await expect(bootstrap_1.bootstrapEnvironment2(env, sdk, {}))
            .rejects.toThrow('Not downgrading existing bootstrap stack');
    });
    afterEach(() => {
        mockDeployStack.mockClear();
    });
});
