// aws-cdk/lib/logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const colors = require("colors/safe");
const util = require("util");
const { stdout, stderr } = process;
const logger = (stream, styles) => (fmt, ...args) => {
    let str = util.format(fmt, ...args);
    if (styles && styles.length) {
        str = styles.reduce((a, style) => style(a), str);
    }
    stream.write(str + '\n');
};
exports.isVerbose = false;
function setVerbose(enabled = true) {
    exports.isVerbose = enabled;
}
exports.setVerbose = setVerbose;
const _debug = logger(stderr, [colors.gray]);
exports.debug = (fmt, ...args) => exports.isVerbose && _debug(fmt, ...args);
exports.error = logger(stderr, [colors.red]);
exports.warning = logger(stderr, [colors.yellow]);
exports.success = logger(stderr, [colors.green]);
exports.highlight = logger(stderr, [colors.bold]);
exports.print = logger(stderr);
exports.data = logger(stdout);
/**
 * Create a logger output that features a constant prefix string.
 *
 * @param prefixString the prefix string to be appended before any log entry.
 * @param fn   the logger function to be used (typically one of the other functions in this module)
 *
 * @returns a new LoggerFunction.
 */
function prefix(prefixString, fn) {
    return (fmt, ...args) => fn(`%s ${fmt}`, prefixString, ...args);
}
exports.prefix = prefix;
