// @aws-cdk/aws-cloudformation/test/integ.core-custom-resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Stack verification steps:
 * - Deploy with `--no-clean`
 * - Verify that the CloudFormation stack outputs have the following values:
 *   - Ref: "MyPhysicalReflectBack"
 *   - GetAtt.Attribute1: "foo"
 *   - GetAtt.Attribute2: 1234
 */
const core_1 = require("@aws-cdk/core");
class TestStack extends core_1.Stack {
    constructor(scope, id) {
        super(scope, id);
        const resourceType = 'Custom::Reflect';
        const serviceToken = core_1.CustomResourceProvider.getOrCreate(this, resourceType, {
            codeDirectory: `${__dirname}/core-custom-resource-provider-fixture`,
            runtime: core_1.CustomResourceProviderRuntime.NODEJS_12,
        });
        const cr = new core_1.CustomResource(this, 'MyResource', {
            resourceType,
            serviceToken,
            properties: {
                physicalResourceId: 'MyPhysicalReflectBack',
                attributes: {
                    Attribute1: 'foo',
                    Attribute2: 1234,
                },
            },
        });
        new core_1.CfnOutput(this, 'Ref', { value: cr.ref });
        new core_1.CfnOutput(this, 'GetAtt.Attribute1', { value: core_1.Token.asString(cr.getAtt('Attribute1')) });
        new core_1.CfnOutput(this, 'GetAtt.Attribute2', { value: core_1.Token.asString(cr.getAtt('Attribute2')) });
    }
}
const app = new core_1.App();
new TestStack(app, 'custom-resource-test');
app.synth();
