// @aws-cdk/aws-amplify/test/domain.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const core_1 = require("@aws-cdk/core");
const amplify = require("../lib");
test('create a domain', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const app = new amplify.App(stack, 'App', {
        sourceCodeProvider: new amplify.GitHubSourceCodeProvider({
            owner: 'aws',
            repository: 'aws-cdk',
            oauthToken: core_1.SecretValue.plainText('secret'),
        }),
    });
    const prodBranch = app.addBranch('master');
    const devBranch = app.addBranch('dev');
    // WHEN
    const domain = app.addDomain('amazon.com', {
        subDomains: [
            {
                branch: prodBranch,
                prefix: 'prod',
            },
        ],
    });
    domain.mapSubDomain(devBranch);
    // THEN
    expect(stack).toHaveResource('AWS::Amplify::Domain', {
        AppId: {
            'Fn::GetAtt': [
                'AppF1B96344',
                'AppId',
            ],
        },
        DomainName: 'amazon.com',
        SubDomainSettings: [
            {
                BranchName: {
                    'Fn::GetAtt': [
                        'Appmaster71597E87',
                        'BranchName',
                    ],
                },
                Prefix: 'prod',
            },
            {
                BranchName: {
                    'Fn::GetAtt': [
                        'AppdevB328DAFC',
                        'BranchName',
                    ],
                },
                Prefix: {
                    'Fn::GetAtt': [
                        'AppdevB328DAFC',
                        'BranchName',
                    ],
                },
            },
        ],
    });
});
test('map a branch to the domain root', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const app = new amplify.App(stack, 'App', {
        sourceCodeProvider: new amplify.GitHubSourceCodeProvider({
            owner: 'aws',
            repository: 'aws-cdk',
            oauthToken: core_1.SecretValue.plainText('secret'),
        }),
    });
    const prodBranch = app.addBranch('master');
    // WHEN
    const domain = app.addDomain('amazon.com');
    domain.mapRoot(prodBranch);
    // THEN
    expect(stack).toHaveResource('AWS::Amplify::Domain', {
        AppId: {
            'Fn::GetAtt': [
                'AppF1B96344',
                'AppId',
            ],
        },
        DomainName: 'amazon.com',
        SubDomainSettings: [
            {
                BranchName: {
                    'Fn::GetAtt': [
                        'Appmaster71597E87',
                        'BranchName',
                    ],
                },
                Prefix: '',
            },
        ],
    });
});
test('throws at synthesis without subdomains', () => {
    // GIVEN
    const app = new core_1.App();
    const stack = new core_1.Stack(app, 'test-stack');
    const amplifyApp = new amplify.App(stack, 'App', {
        sourceCodeProvider: new amplify.GitHubSourceCodeProvider({
            owner: 'aws',
            repository: 'aws-cdk',
            oauthToken: core_1.SecretValue.plainText('secret'),
        }),
    });
    // WHEN
    amplifyApp.addDomain('amazon.com');
    // THEN
    expect(() => app.synth()).toThrow(/The domain doesn't contain any subdomains/);
});
