// @aws-cdk/aws-dynamodb/test/integ.dynamodb.mixed-key-gsi.js
"use strict";
/**
 * This aimes to verify we can deploy a DynamoDB table with an attribute being
 * a key attribute in one GSI, and a non-key attribute in another.
 *
 * See https://github.com/aws/aws-cdk/issues/4398
 */
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
const app = new core_1.App();
const stack = new core_1.Stack(app, 'aws-cdk-dynamodb-gsi');
const table = new lib_1.Table(stack, 'Table', {
    partitionKey: { name: 'pkey', type: lib_1.AttributeType.NUMBER },
    removalPolicy: core_1.RemovalPolicy.DESTROY,
});
table.addGlobalSecondaryIndex({
    indexName: 'IndexA',
    partitionKey: { name: 'foo', type: lib_1.AttributeType.STRING },
    projectionType: lib_1.ProjectionType.INCLUDE,
    nonKeyAttributes: ['bar'],
});
table.addGlobalSecondaryIndex({
    indexName: 'IndexB',
    partitionKey: { name: 'baz', type: lib_1.AttributeType.STRING },
    sortKey: { name: 'bar', type: lib_1.AttributeType.STRING },
    projectionType: lib_1.ProjectionType.INCLUDE,
    nonKeyAttributes: ['blah'],
});
