// cdk-assets/test/mock-aws.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
jest.mock('aws-sdk');
const AWS = require("aws-sdk");
function mockAws() {
    const mockEcr = new AWS.ECR();
    const mockS3 = new AWS.S3();
    // Sane defaults which can be overridden
    mockS3.getBucketLocation = mockedApiResult({});
    mockEcr.describeRepositories = mockedApiResult({ repositories: [
            {
                repositoryUri: '12345.amazonaws.com/repo',
            },
        ] });
    return {
        mockEcr,
        mockS3,
        discoverCurrentAccount: jest.fn(() => Promise.resolve({ accountId: 'current_account', partition: 'swa' })),
        discoverDefaultRegion: jest.fn(() => Promise.resolve('current_region')),
        ecrClient: jest.fn().mockReturnValue(Promise.resolve(mockEcr)),
        s3Client: jest.fn().mockReturnValue(Promise.resolve(mockS3)),
    };
}
exports.mockAws = mockAws;
function errorWithCode(code, message) {
    const ret = new Error(message);
    ret.code = code;
    return ret;
}
exports.errorWithCode = errorWithCode;
function mockedApiResult(returnValue) {
    return jest.fn().mockReturnValue({
        promise: jest.fn().mockResolvedValue(returnValue),
    });
}
exports.mockedApiResult = mockedApiResult;
function mockedApiFailure(code, message) {
    return jest.fn().mockReturnValue({
        promise: jest.fn().mockRejectedValue(errorWithCode(code, message)),
    });
}
exports.mockedApiFailure = mockedApiFailure;
/**
 * Mock upload, draining the stream that we get before returning
 * so no race conditions happen with the uninstallation of mock-fs.
 */
function mockUpload(expectContent) {
    return jest.fn().mockImplementation(request => ({
        promise: () => new Promise((ok, ko) => {
            const didRead = new Array();
            const bodyStream = request.Body;
            bodyStream.on('data', (chunk) => { didRead.push(chunk.toString()); }); // This listener must exist
            bodyStream.on('error', ko);
            bodyStream.on('close', () => {
                const actualContent = didRead.join('');
                if (expectContent !== undefined && expectContent !== actualContent) {
                    throw new Error(`Expected to read '${expectContent}' but read: '${actualContent}'`);
                }
                ok();
            });
        }),
    }));
}
exports.mockUpload = mockUpload;
