// @aws-cdk/aws-sam/test/sam.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const cdk = require("@aws-cdk/core");
const lib_1 = require("../lib");
test('construct an AWS::Serverless::Application', () => {
    const stack = new cdk.Stack();
    new lib_1.CfnApplication(stack, 'App', {
        location: {
            applicationId: 'arn:aws:serverlessrepo:us-east-1:077246666028:applications/aws-serverless-twitter-event-source',
            semanticVersion: '2.0.0',
        },
        parameters: {
            SearchText: '#serverless -filter:nativeretweets',
            TweetProcessorFunctionName: 'test',
        },
    });
    expect(stack).toMatchTemplate({
        Transform: 'AWS::Serverless-2016-10-31',
        Resources: {
            App: {
                Type: 'AWS::Serverless::Application',
                Properties: {
                    Location: {
                        ApplicationId: 'arn:aws:serverlessrepo:us-east-1:077246666028:applications/aws-serverless-twitter-event-source',
                        SemanticVersion: '2.0.0',
                    },
                    Parameters: {
                        SearchText: '#serverless -filter:nativeretweets',
                        TweetProcessorFunctionName: 'test',
                    },
                },
            },
        },
    });
});
